### ISSUE TITLE
Duplicate app icons show in MacOS Dock

### ISSUE THREAD

**Issue by mattbierner** [microsoft/vscode#138475]
Issue Type: <b>Bug</b>

After starting VS Code, I see duplicate bouncing icons in the MacOS dock

There is one new icon for each workspace I have opened. They never resolve and are always bouncing

VS Code version: Code - Insiders 1.63.0-insider (d09289a2b6dd0e56b2a264a8e9380685498e10d6, 2021-12-04T02:02:08.723Z)
OS version: Darwin x64 21.1.0
Restricted Mode: No

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|Intel(R) Core(TM) i7-7700K CPU @ 4.20GHz (8 x 4200)|
|GPU Status|2d_canvas: enabled<br>gpu_compositing: enabled<br>metal: disabled_off<br>multiple_raster_threads: enabled_on<br>oop_rasterization: enabled<br>opengl: enabled_on<br>rasterization: enabled<br>skia_renderer: disabled_off_ok<br>video_decode: enabled<br>webgl: enabled<br>webgl2: enabled|
|Load (avg)|3, 9, 7|
|Memory (System)|40.00GB (2.32GB free)|
|Process Argv|--crash-reporter-id cefe23df-690f-4b1b-9ce9-e004d3be2745|
|Screen Reader|no|
|VM|0%|
</details>

**Comment by mattbierner** [comment]
Ah looks like  ban.spellright is somehow causing this: https://github.com/bartosz-antosik/vscode-spellright/issues/463

Not even sure how that's possible and the extension was updated on 8/10/2021. I definitely was using VS Code insiders after that update so I wonder if something has changed on our side that triggered this

**Comment by deepak1556** [comment]
Can you provide some steps to repro the issue, I tried the following on macOS Monterey and don't see the above duplicate icons.

1) Disable `Show recent application in Dock` from dock preferences
2) Install spellright extension
3) `code-insiders workspace1`
4) `code-insiders workspace2`
5) `code-insiders workspace3`

**Comment by mattbierner** [comment]
Sorry no additional repo steps that I know. I confirmed I have `Show recent application in Dock` disabled

Enabling the extension immediately triggers the bug. Are you are aware of any way to investigate which specific process is associated with the dock icons?

**Comment by mjbvz** [comment]
Also seeing this on my work machine

**Comment by deepak1556** [comment]
> Are you are aware of any way to investigate which specific process is associated with the dock icons?

Yup that would be the main process, and also we don't call any native apis via the runtime that would create the dock icon instead the flow today works like this,

When the first instance of the application is launched the runtime will add the instance to the dock and in this instance we run a server in the main process that creates a socket in the user data dir. Now when newer instances of the application are launched, basically step 4) and 5) from https://github.com/microsoft/vscode/issues/138475#issuecomment-986649907 these instances will first check for the above socket presence and if it is then they will pass the arguments, state related to this launch to the first instance via the socket and kill themselves which is what a user perceives as a brief appearance of a new icon in the dock.

But there are situations in which we explicitly allow for new instances to have their normal lifecycle instead of following the above singleton lifecycle. These are when you launch the cli with either the `--wait` flag or when you launch with `-` which allows you to read from stdin.

So, coming back to the issue at hand, it is interesting an extension is able to impact the above flow given they cannot control the main process related behavior. But what we can first confirm is, with what flags these instances are launched in your scenario

1) Launch the instances from the cli as you normally do with just one additional argument prepended `NODE_DEBUG=child_process` and also `--verbose` to get the output

**Comment by mjbvz** [comment]
Thanks. I followed those instructions and here are the log files: 

[20211206T121230.zip](https://github.com/microsoft/vscode/files/7663304/20211206T121230.zip)

**Comment by ddbrierton** [comment]
Seeing this also with the newly released VS Code 1.63.0 on macOS Big Sur 11.6.1, also using the spellright extension.
